core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -813,7 +813,14 @@ private[spark] class BlockManager(
               false
           }
         } else {
-          memoryStore.putBytes(blockId, size, level.memoryMode, () => bytes)
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP) {
```

Comment:
NM about checking `memoryMode.useOffHeap`. I got that confused with `StorageLevel`. There's actually only two values of `MemoryMode`: `MemoryMode.OFF_HEAP` and `MemoryMode.ON_HEAP`.